### PR TITLE
Null safety

### PR DESCRIPTION
**Pull request by markst**
First progress of null safety migration. 

Opening up for feedback. 

#1

### CODE REVIEW

File: pubspec.yaml [hunk 1]
```diff
@@ -5,23 +5,21 @@ homepage: https://www.peterscode.dev
...
-  upnp: ^2.0.1
-  dart_chromecast: ^0.2.4
-  multicast_dns: ^0.2.2
-  xml: '>=4.0.0 <5.0.0'
-  file: '>=5.0.0 <6.0.0'
-  js: ^0.6.2
-  #upnp:
-  #  git: https://github.com/peterscodee/upnp_dart.git
+  http: ^0.13.4
+  http_server: ^1.0.0
```
Comment (markst) [inline pubspec.yaml:14]:
```bash
audio_cast % dart pub get    
Resolving dependencies... 
  http_server 1.0.0 (discontinued replaced by shelf)
  pedantic 1.11.1 (discontinued replaced by lints)
```

**Comment by daniel-naegele** [comment]
I would propose to use my version of the original `upnp` fork [upnp2](https://pub.dev/packages/upnp2), because I want to actively maintain, improve, document, and add tests to this `upnp` fork.

File: lib/src/adapter/upnp_adapter.dart [hunk 2]
```diff
@@ -31,7 +31,9 @@ class UPnPAdapter extends CastAdapter {
     await disc.start(ipv6: false);
```
Comment (markst) [inline lib/src/adapter/upnp_adapter.dart:31]: @Butzlabben happy to, but does seem to be raising cast error exception:

```
_CastError (type '() => Null' is not a subtype of type '(() => FutureOr<HttpClientRequest>)?' in type cast)
```

https://github.com/Butzlabben/upnp.dart/blob/main/lib/src/service.dart#L57

**Comment by markst** [comment]
@Butzlabben happy to, but does seem to be raising cast error exception:

```
_CastError (type '() => Null' is not a subtype of type '(() => FutureOr<HttpClientRequest>)?' in type cast)
```

https://github.com/Butzlabben/upnp.dart/blob/main/lib/src/service.dart#L57

**Comment by daniel-naegele** [comment]
@markst I think I got it fixed now in 3.0.5.

This was a result from the null safety migration from the original pull request to the original repository

**Comment by dxvid-pts** [comment]
> I would propose to use my version of the original `upnp` fork [upnp2](https://pub.dev/packages/upnp2), because I want to actively maintain, improve, document, and add tests to this `upnp` fork.

I remember that something was not working with the original upnp package. Therefore I had to use modified code (see https://github.com/peterscodee/upnp_dart), which I copied into an extension: https://github.com/peterscodee/audio_cast/blob/e44c4235510caef14b9ec1311cd4577ba4399403/lib/src/adapter/upnp_adapter.dart#L229

Do you know anything about that?

Also I will need some time to review this. The main difficulty is my current lack of test devices. Can other people confirm this is not breaking on their setup?

**Comment by daniel-naegele** [comment]
So I looked into the code, an interestingly, the method in the `EditedAction` does never actually get used (at least according to IntelliJ). Also, the only real changes in your fork from the original upnp is [this commit ](https://github.com/peterscodee/upnp_dart/commit/cdf28f490e3a4a9fed39e540b8aeeb6d87028ac9). 

[In the specification of upnp](https://openconnectivity.org/upnp-specs/UPnP-arch-DeviceArchitecture-v2.0-20200417.pdf) it says (page 79 at the bottom)
```
When the value of any argument contains one or more characters reserved as markup (such as ampersand (“&”) or less than (“<”)), then the text shall be escaped in accordance with the provisions of clause 2.4 of the XML specification and each such character replaced with the equivalent numeric representation or string (such as “&amp;” or “&lt;”). Such characters appearing in URLs are also allowed to be percent-encoded in accordance with the URL percent-encoding rules specified in clauses 2.1 and 2.4 of RFC 3986.
```

According to [this](https://docs.oracle.com/cd/A97335_02/apps.102/bc4j/developing_bc_projects/obcCustomXml.htm) or [this](https://www.w3resource.com/xml/reserved-markup-characters.php), I will change the code to this:
```dart
final param = '  <u:$name xmlns:u="${service.type}">' +
        args.keys.map((argumentName) {
          String argument = args[argumentName].toString();
          argument = argument.replaceAll('&', '&amp;');
          argument = argument.replaceAll('<', '&lt;');
          argument = argument.replaceAll('>', '&gt;');
          argument = argument.replaceAll('\'', '&apos;');
          argument = argument.replaceAll('\"', '&qout;');

          return '<$argumentName>$argument</$argumentName>';
        }).join('\n') +
        '</u:$name>\n';
```

Any other feedback/suggestions?

**Comment by daniel-naegele** [comment]
@peterscodee Also note, that after this PR you could set the package back to listed on `pub.dev`, as it is currently unlisted

**Comment by dxvid-pts** [comment]
@Butzlabben I just gave you collaborator rights, so my (current) response time doesn't block you from contributing.

I would wait 1 - 2 months before publishing a stable version on pub.dev.
I want to make sure everything works as expected and is testet appropriately.

**Comment by markst** [comment]
> @markst I think I got it fixed now in 3.0.5.
> 
> This was a result from the null safety migration from the original pull request to the original repository

Thanks @Butzlabben.

Unfortunately away at the moment and don't have access to a Chromecast device.

File: pubspec.yaml (hunk 1 above)
Comment (daniel-naegele) [reply pubspec.yaml:14]: I would make a seperate PR for replacing these discontinued dependencies

**Review by daniel-naegele — APPROVED** [review]
I guess this Pull Request is quite OK as is. But before a release to pub.dev, a lot of refactoring will be needed to make code cleaner, safer and test it more. 
@peterscodee I would merge this PR in a week or so, if you have nothing against it.

**Pull request merged**